src/network/networkd-json.c

```diff
@@ -1161,6 +1162,141 @@ static int dns_misc_build_json(Link *link, JsonVariant **ret) {
         return r;
 }
 
+static int dhcp_server_offered_leases_build_json(Link *link, JsonVariant **ret) {
+        _cleanup_(json_variant_unrefp) JsonVariant *elements = NULL;
+        int r;
+        DHCPLease *lease;
+
+        assert(link);
+        assert(ret);
+
+        if (!link->dhcp_server) {
+                *ret = NULL;
+                return 0;
+        }
+
+        if (hashmap_isempty(link->dhcp_server->bound_leases_by_client_id)) {
+                *ret = NULL;
+                return 0;
+        }
+
+        HASHMAP_FOREACH(lease, link->dhcp_server->bound_leases_by_client_id) {
+                _cleanup_(json_variant_unrefp) JsonVariant *e = NULL;
+                _cleanup_free_ char *id = NULL;
+                struct in_addr address = { .s_addr = lease->address };
+
+                r = sd_dhcp_client_id_to_string(lease->client_id.data, lease->client_id.length, &id);
+                if (r < 0) {
+                        *ret = NULL;
+                        return 0;
+                }
+
+                r = json_build(&e,JSON_BUILD_OBJECT(
+                        JSON_BUILD_PAIR_STRING("ClientId", id),
+                        JSON_BUILD_PAIR_IN4_ADDR_NON_NULL("Address", &address),
+                        JSON_BUILD_PAIR_UNSIGNED_NON_ZERO("Expiration", lease->expiration)));
+                if (r < 0) {
+                        *ret = NULL;
+                        return 0;
+                }
+
+                r = json_variant_append_array(&elements, e);
+                if (r < 0) {
+                        *ret = NULL;
+                        return 0;
+                }
+        }
+
+        return json_build(ret, JSON_BUILD_OBJECT(JSON_BUILD_PAIR_VARIANT("Leases", elements)));
+}
+
+static int dhcp_server_static_leases_build_json(Link *link, JsonVariant **ret) {
+        _cleanup_(json_variant_unrefp) JsonVariant *elements = NULL;
+        int r;
+        DHCPLease *lease;
+
+        assert(link);
+        assert(ret);
+
+        if (!link->dhcp_server) {
+                *ret = NULL;
+                return 0;
+        }
+
+        if (hashmap_isempty(link->dhcp_server->static_leases_by_client_id)) {
+                *ret = NULL;
+                return 0;
+        }
+
+        HASHMAP_FOREACH(lease, link->dhcp_server->static_leases_by_client_id) {
+                _cleanup_(json_variant_unrefp) JsonVariant *e = NULL;
+                _cleanup_free_ char *id = NULL;
+                struct in_addr address = { .s_addr = lease->address };
+
+                r = sd_dhcp_client_id_to_string(lease->client_id.data, lease->client_id.length, &id);
```

Comment:
Also here.
And, currently, we support only MAC address to be matched.
Hence, please check the first byte and drop the first byte from the json data.

Reply:
So is your suggestion
- for offered leases to return the client id as is in a field called "ClientId" and leave it up to the consumer of the JSON to decode further
- and for static leases to return the MACAddress/HardwareAddress

I would rather keep it the same for both lists of leases to make it easier for whoever uses this data. Especially since the ClientId for static leases in systemd-networkd only being able to be MAC Addresses is an implementation detail that holds true today, but might not be so true in the future, requiring more specific logic to expose this information.

Maybe a middle group is to always return the `ClientId`, and then in extra fields decode the ClientId into whatever it is and return it as separate fields?

Reply:
The revised version looks good to me.